source/connection_manager.c

```diff
@@ -756,6 +763,133 @@ static void s_schedule_connection_culling(struct aws_http_connection_manager *ma
+                /* Use Tunneling when main connection use TLS. */
+                proxy_options->connection_type = AWS_HPCT_HTTP_TUNNEL;
+            } else {
+                /* Use forwarding proxy when main connection use clear text. */
+                proxy_options->connection_type = AWS_HPCT_HTTP_FORWARD;
+            }
+        }
+        if (aws_byte_cursor_eq_ignore_case(&proxy_uri.scheme, &aws_http_scheme_https)) {
+            if (options->proxy_env_var_settings.tls_options) {
+                proxy_options->tls_options = options->proxy_env_var_settings.tls_options;
+            } else {
+                struct aws_tls_ctx *tls_ctx = NULL;
+                struct aws_tls_ctx_options tls_ctx_options;
+                AWS_ZERO_STRUCT(tls_ctx_options);
+                /* create a default tls options */
+                aws_tls_ctx_options_init_default_client(&tls_ctx_options, allocator);
+                /* turn off verify peer for proxy connection setup by default */
+                aws_tls_ctx_options_set_verify_peer(&tls_ctx_options, false);
+                tls_ctx = aws_tls_client_ctx_new(allocator, &tls_ctx_options);
+                aws_tls_ctx_options_clean_up(&tls_ctx_options);
+                if (!tls_ctx) {
+                    AWS_LOGF_ERROR(AWS_LS_HTTP_CONNECTION_MANAGER, "Failed to create default TLS context.");
+                    goto done;
+                }
+                aws_tls_connection_options_init_from_ctx(&default_tls_connection_options, tls_ctx);
+                /* tls options hold a ref to the ctx */
+                aws_tls_ctx_release(tls_ctx);
+                if (aws_tls_connection_options_set_server_name(
+                        &default_tls_connection_options, allocator, &proxy_uri.host_name)) {
+                    AWS_LOGF_ERROR(
+                        AWS_LS_HTTP_CONNECTION_MANAGER, "Failed set server name for TLS connection options.");
+                    goto done;
+                }
+                proxy_options->tls_options = &default_tls_connection_options;
+            }
+        }
+        /* Support basic authentication. */
+        if (proxy_uri.password.len) {
+            /* Has no empty password set */
+            struct aws_http_proxy_strategy_basic_auth_options config = {
+                .proxy_connection_type = proxy_options->connection_type,
+                .user_name = proxy_uri.user,
+                .password = proxy_uri.password,
+            };
+            struct aws_http_proxy_strategy *proxy_strategy = aws_http_proxy_strategy_new_basic_auth(allocator, &config);
+            proxy_options->proxy_strategy = proxy_strategy;
+        }
+        manager->proxy_config = aws_http_proxy_config_new_from_proxy_options(allocator, proxy_options);
+        /* config has the deep copy of the needed options, we can clean up resources right after it */
+        aws_tls_connection_options_clean_up(&default_tls_connection_options);
```

Comment:
seems like all of these need to be move up to the top of the function and cleaned up unconditionally in done